### ISSUE TITLE
When pruning images, Skaffold removes images that were created by Docker Compose

### ISSUE THREAD

**Issue by jpetazzo** [GoogleContainerTools/skaffold#2464]
### Expected behavior

Skaffold leaves my existing images alone.

(To clarify, I'm talking about images stored in my local Docker Engine.)

### Actual behavior

Skaffold removes images that I've built with Docker Compose. This defeats the build cache that I use with my other tools.

Note: it doesn't remove images that I've built manually (with `docker build`) but it does remove images that I've built with `docker-compose build`.

### Information

- Skaffold version: v0.33.0
- Operating system: Linux
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v1beta12
kind: Config
build:
  artifacts:
  - image: localhost:30000/hasher
    context: hasher
  - image: localhost:30000/rng
    context: rng
  - image: localhost:30000/webui
    context: webui
  - image: localhost:30000/worker
    context: worker
deploy:
  kubectl:
    manifests:
    - kubercoins/hasher-deployment.yaml
    - kubercoins/hasher-service.yaml
    - kubercoins/redis-deployment.yaml
    - kubercoins/redis-service.yaml
    - kubercoins/rng-daemonset.yaml
    - kubercoins/rng-service.yaml
    - kubercoins/webui-deployment.yaml
    - kubercoins/webui-service.yaml
    - kubercoins/worker-deployment.yaml
```
(This file was created by `skaffold init`.)

### Steps to reproduce the behavior

1. Use Compose for local development.
2. Run `skaffold dev`.
3. Wait for Skaffold to build+run all the things.
4. Stop Skaffold (I used Ctrl-C).
5. Watch as Skaffold prunes its images.
6. Check with `docker images` that the images previously built by Compose are gone.

I suppose I can work around the bug by running `skaffold dev --no-prune` or setting the env var `SKAFFOLD_NO_PRUNE`. Or maybe there is also a way to do that in the Skaffold configuration file?

By the way, what's the rationale for pruning the images? I couldn't find an explanation why.

Thank you!

**Comment by dgageot** [comment]
Thank you @jpetazzo for opening this issue. The rationale for pruning is that skaffold produces a lot of images in dev mode. That ends up filling the user's drive.
However, I think the current implementation of pruning is super sub optimal.
ping @nkubala

**Comment by japsen** [comment]
@jpetazzo a workaround might be having a "base image" which contains installed packages, updates, etc.  and using a separate Dockerfile in your skaffold.yaml which only adds the latest code and/or does the really necessary things.
So just something like this:
```
FROM eu.gcr.io/reponame/baseimagename
ADD ./folder /destination
```
 I'm using it like this and skaffold doesn't prune my base images.

**Comment by balopat** [comment]
Hi @jpetazzo - Besides trying to optimize prune, I could see this included in the global skaffold config as a persistent user preference. What do you think?

**Comment by jpetazzo** [comment]
@japsen Thanks for the suggestion; unfortunately, that's a bit too disruptive for me, as it makes my local development workflow way more complicated (but perhaps I can get some inspiration from that, so thanks!)

Apparently, if I manually tag the image, Skaffold doesn't remove it. (I will have to retry later, but when I tried to reproduce the issue by manually tag the image, Skaffold didn't remove it; however when I reproduced again with Compose, it did remove it.) 

@balopat (Keeping in mind that I'm more of a backend person, and user experience / developer experience is not my forte) Sure, having a global setting could help. Personally I would probably disable pruning altogether, and just use "docker system prune" once in a while. For me, the ideal behavior would be to prune old images but keep the one I just built.

Thank you!

**Comment by nkubala** [comment]
regardless of the UX behind disabling skaffold's pruning by default, this behavior shouldn't be happening. I think because I was originally seeing hundreds of images cluttering my docker daemon after running skaffold many times per day, my initial approach to getting rid of the images skaffold builds was too aggressive, but skaffold definitely should not be removing images that it didn't build directly.

@jpetazzo are the images you're building with compose being used directly as base images for the ones you're building with skaffold?

**Comment by jpetazzo** [comment]
OK, here is a tiny example to reproduce and illustrate the issue.

First, create a directory with the following files:
```
mkdir -p /tmp/skaffold-2464 && cd /tmp/skaffold-2464

cat >docker-compose.yml <<EOF
version: "3"
services:
  hoptopod:
    build: .
EOF

cat >Dockerfile <<EOF
FROM alpine
RUN sleep 3
EOF

cat >pod.yml <<EOF
kind: Pod
apiVersion: v1
metadata:
  name: hoptopod
spec:
  containers:
  - name: hoptopod
    image: localhost:5000/hoptopod
EOF
```

Then start a local registry:
```
docker run -d -p 5000:5000 registry
```

Bring up the app with Compose:
```
docker-compose up
```
(The container will exit immediately, that's normal.)

Bring up the app with Skaffold:
```
skaffold dev
```
(It will push the app to the local registry; then possibly deploy it to your k8s cluster; then exit and prune images.)

Re-run with Compose:
```
docker-compose up
```
(It will rebuild the image, including the "expensive" `RUN sleep 3` layer. Furthermore, it triggers an interesting Compose issue, as it notices that the image has been removed.)

**Comment by tvvignesh** [comment]
@dgageot You are right. Loads of images created by skaffold in my local machine. Would be great to have just the latest image built on the GIT tag cached.

![Screenshot from 2019-12-27 16-42-15](https://user-images.githubusercontent.com/1165845/71515145-e099cc00-28c7-11ea-88cd-35848cb69342.png)

**Comment by tstromberg** [comment]
Perhaps I am missing something, but this seems like incorrect behavior to me - shouldn't Skaffold only prune images created by Skaffold?

**Comment by tstromberg** [comment]
Can someone provide `-v trace` logs for this? It would go a long way to figuring out what calls we are sending to Docker in this case.

There may be some issue with Docker pruning children with interdependent layers, but we don't yet have the evidence yet for this.

**Comment by briandealwis** [comment]
@tvvignesh or @jpetazzo — would you be able to run `skaffold dev -v trace`?

**Comment by tvvignesh** [comment]
@briandealwis Oops! I changed my workflow since I had raised this issue (I use remote clusters for development and with skaffold the image has to be built everytime and pushed to remote registry causing the approach to be unsustainable). Now, I use telepresence locally to proxy during development and use Skaffold in CI/CD pipelines and thus, I don't have any such images locally to reproduce and test this now.
